kvserver: add onProcessResult and onEnqueueResult to processCallback

wenyihu6 opened the pull request:
Part of: https://github.com/cockroachdb/cockroach/issues/151847
Epic: none 

---

**kvserver: use non-blocking send on errors for maybeBackpressureBatch**

maybeBackpressureBatch registers a callback with the split queue for replicas
that are too large relative to their split size. This backpressures the range to
stop it from growing and prevent new writes from blocking a pending split. The
callback is invoked when the split queue finishes processing the replica.

Previously, the error channel used in the callback had a size of 1 and performed
blocking sends. This was safe because the base queue only sent a single error,
and by the time maybeBackpressureBatch returned, the callback was guaranteed to
have been consumed, and no additional sends would occur.

Future commits will allow the callback to be invoked multiple times (although it
should only twice at most). To be safe and avoid potential deadlocks from
multiple sends after maybeBackpressureBatch already returns, this commit makes
the error send non-blocking. If the channel is already full, the error is
dropped, which is acceptable since we only care about observing the completion
of the replica processing at least once.

---

**kvserver: return baseQueueAsyncRateLimited from bq.Async**

baseQueue.Async may return immediately as a noop if the semaphore does not
available capacity and the wait parameter is false. Previously, this case
returned no error, leaving the caller unaware that the request was dropped. This
commit changes the behavior to return a baseQueueAsyncRateLimited error,
allowing callers to detect and handle the condition.

---

**kvserver: add onProcessResult and onEnqueueResult to processCallback**

The base queue already supports registering callbacks that are invoked with the
processing result of replicas once they are processed. However, replicas may
fail before reaching that stage (e.g., failing to enqueue or dropped early).
This commit extends the mechanism to also report enqueue results, allowing
callers to detect failures earlier. Currently, only
decommissioningNudger.maybeEnqueueProblemRange uses this.

Note that one behavior change is introduced: previously, a registered callback
would fire only once with the processing result and not again if the replica was
later processed by the purgatory queue. With this change, the callback may now
be invoked twice.

---

**kvserver: add TestBaseQueueCallback**

This commit adds TestBaseQueueCallbackOnEnqueueResult and
TestBaseQueueCallbackOnProcessResult to verify that callbacks are correctly
invoked with both enqueue and process results.

wenyihu6 committed:
kvserver: add onProcessResult and onEnqueueResult to processCallback

The base queue already supports registering callbacks that are invoked with the
processing result of replicas once they are processed. However, replicas may
fail before reaching that stage (e.g., failing to enqueue or dropped early).
This commit extends the mechanism to also report enqueue results, allowing
callers to detect failures earlier. Currently, only
decommissioningNudger.maybeEnqueueProblemRange uses this.

Note that one behavior change is introduced: previously, a registered callback
would fire only once with the processing result and not again if the replica was
later processed by the purgatory queue. With this change, the callback may now
be invoked twice.

wenyihu6 committed:
kvserver: return baseQueueAsyncRateLimited from bq.Async

baseQueue.Async may return immediately as a noop if the semaphore does not
available capacity and the wait parameter is false. Previously, this case
returned no error, leaving the caller unaware that the request was dropped. This
commit changes the behavior to return a baseQueueAsyncRateLimited error,
allowing callers to detect and handle the condition.

wenyihu6 committed:
kvserver: add TestBaseQueueCallback

This commit adds TestBaseQueueCallbackOnEnqueueResult and
TestBaseQueueCallbackOnProcessResult to verify that callbacks are correctly
invoked with both enqueue and process results.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152792)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver: use non-blocking send on errors for maybeBackpressureBatch

maybeBackpressureBatch registers a callback with the split queue for replicas
that are too large relative to their split size. This backpressures the range to
stop it from growing and prevent new writes from blocking a pending split. The
callback is invoked when the split queue finishes processing the replica.

Previously, the error channel used in the callback had a size of 1 and performed
blocking sends. This was safe because the base queue only sent a single error,
and by the time maybeBackpressureBatch returned, the callback was guaranteed to
have been consumed, and no additional sends would occur.

Future commits will allow the callback to be invoked multiple times (although it
should only twice at most). To be safe and avoid potential deadlocks from
multiple sends after maybeBackpressureBatch already returns, this commit makes
the error send non-blocking. If the channel is already full, the error is
dropped, which is acceptable since we only care about observing the completion
of the replica processing at least once.

tbg reviewed (changes requested):
Nice, looking good! Some comments about de-risking the logging while keeping it useful (and misc other comments, none of them of particular substance).

> @tbg reviewed 1 of 1 files at r1, 3 of 3 files at r2, 5 of 5 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-:-OZ3iXXl8NfMllevfr0J:b-wraj6v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/replica_backpressure.go` line 214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3eIn40VDZVSLZ0oLp:-OZ3eIn40VDZVSLZ0oLq:b88en8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa26dc3b2812e6af8e8c1e418d3a011082278d7c/pkg/kv/kvserver/replica_backpressure.go#L214)):*
> ```Go
> 			case splitC <- err:
> 			default:
> 				// TODO(wenyihu6): should we add ctx timeout when invoking callbacks
> ```

this TODO seems out of place, and along with the unrelated comment below it tripped me up a little. Consider removing or moving the TODO.
___
*[`pkg/kv/kvserver/queue.go` line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3ejgs2Mp45HPjQFGD:-OZ3ejgs2Mp45HPjQFGE:bpxwuts) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L132)):*
> ```Go
> // before processing due to exceeding max queue size.
> // 2. onProcessResult may be called with error first and sent to the purgatory
> // queue and called again when the puragtory processes the replica.
> ```

puragtory
___
*[`pkg/kv/kvserver/queue.go` line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3esba1Q4PzoFwEi_M:-OZ3esba1Q4PzoFwEi_N:b-f4tv9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L150)):*
> ```Go
> 	onProcessResult func(err error)
> 
> 	// onEnqueueResult is called with the result of the enqueue attempt. It is
> ```

nit: for temporal consistency, move `onEnqueue` up above `onProcessResult`.
___
*[`pkg/kv/kvserver/queue.go` line 163 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3f1iQ8eAMwiF6Hnfn:-OZ3f1iQ8eAMwiF6Hnfo:bs6y1rs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L163)):*
> ```Go
> 	// callback is -1. Note: indexOnHeap does not represent the item's exact rank
> 	// by priority. It only reflects the item's position in the heap array, which
> 	// gives a rough idea of where it sits in the priority hierarchy.
> ```

Here too could you reflect on whether the callback is necessarily invoked at all? I imagine there are cases where it isn't, and it'd be good to call this out.
___
*[`pkg/kv/kvserver/queue.go` line 645 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3fGAbC_JlrJ3oghio:-OZ3fGAbC_JlrJ3oghip:b-vq89ne) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L645)):*
> ```Go
> type queueHelper interface {
> 	MaybeAdd(ctx context.Context, repl replicaInQueue, now hlc.ClockTimestamp)
> 	Add(ctx context.Context, repl replicaInQueue, prio float64, processCallback processCallback)
> ```

the new variable name shadows the type, which isn't ideal. Maybe just `processCB processCallback`.
___
*[`pkg/kv/kvserver/queue.go` line 703 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3fMWuCWl63vidzMdk:-OZ3fMWuCWl63vidzMdl:buh13t5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L703)):*
> ```Go
> // enqueued or processed.
> func (bq *baseQueue) AddAsyncWithCallback(
> 	ctx context.Context, repl replicaInQueue, prio float64, processCallback processCallback,
> ```

ditto (maybe just search for `processCallback processCallback` to get all the offenders.
___
*[`pkg/kv/kvserver/queue.go` line 800 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3fWO28BKMvBO34BB_:-OZ3fWO28BKMvBO34BBa:bn7uto0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L800)):*
> ```Go
> 		// We checked this above in MaybeAdd(), but we need to check it
> 		// again for Add().
> 		processCallback.onEnqueueResult(-1 /*indexOnHeap*/, errReplicaNotInitialized)
> ```

It would be more robust to do named returns and then add a defer:

```go
func (bq *baseQueue) addInternal(/* ... */) (_ bool, err error) {
defer func() {
  if err != nil {
    processCB.onEnqueueResult(-1 /* indexOnHeap */, err)
  }
}()
}
```
___
*[`pkg/kv/kvserver/queue.go` line 849 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3g71QD9xK3OUfxS5U:-OZ3g71QD9xK3OUfxS5V:b-5qws05) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L849)):*
> ```Go
> 				log.Dev.Infof(ctx, "updating priority: %0.3f -> %0.3f", item.priority, priority)
> 			}
> 			// TODO(wenyihu6): will this introduce a lot of new memory allocation?
> ```

Would it be better to treat this as a success case? After all, the caller will most likely care about whether the replica is now in the queue, and it is.
___
*[`pkg/kv/kvserver/queue.go` line 874 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3gT5N02Jlv7l1ohKw:-OZ3gT5N02Jlv7l1ohKx:b-yxt14f) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L874)):*
> ```Go
> 			priority, replicaItemToDrop.replicaID)
> 		// TODO(wenyihu6): when we introduce base queue max size cluster setting,
> 		// remember to invoke this callback when shrinking the size
> ```

Adding an extra anchor comment here as a reminder to do this, since I reviewed the shrinking size code already. Likely best to stack this PR on top.

There is also the case in `SetQueueSize` that would need to call callbacks, so most likely a good idea to factor out a single method that is called from both places to do the callback handling.
___
*[`pkg/kv/kvserver/replica.go` line 2968 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3gsDpCBTq6t5y5AIG:-OZ3gsDpCBTq6t5y5AIH:b-74b298) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/replica.go#L2968)):*
> ```Go
> 					// this function to another file so that we can avoid the spam on
> 					// other logs.
> 					log.KvDistribution.Infof(ctx,
> ```

Some rate limiting here would be good here:

var logBudgetEnqueueResult atomic.Int
logBudgetEnqueueResult.Store(15)

// in the callback:
if logBudgetEnqueueResult.Add(-1) > 0 {
  // log.Infof
} else {
  // vmodule
}

So basically each decommission nudger run will produce at most 15 messages for enqueue failures (I didn't name the var very well: it was meant to be for the failure case, we should have one atomic per message type).

Alternatively, and simpler, we could gate all logging behind verbosity if the number of replicas we're nudging is (say) >15. But I like getting some information even in the case in which there are more than that number of replicas, which this atomic-based thing would give us.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152792)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/queue.go` line 800 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3fWO28BKMvBO34BB_:-OZ71lm44FVAWSvvDkAd:bkfpbqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L800)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It would be more robust to do named returns and then add a defer:

```go
func (bq *baseQueue) addInternal(/* ... */) (_ bool, err error) {
defer func() {
  if err != nil {
    processCB.onEnqueueResult(-1 /* indexOnHeap */, err)
  }
}()
}
```
</blockquote></details>

I’m calling `onEnqueueResult` with an error even when the returned error is `nil` in certain cases, such as `errReplicaAlreadyInPurgatory` or `errReplicaAlreadyProcessing`. I want to discuss the desired behavior here before making the changes, especially in light of your point below about whether a replica already in the queue should be treated as a success. 

I still want visibility into these cases - at least knowing if a replica is somehow stuck in purgatory or processing. While the replicas are technically in the queue and things might seem fine, I’m concerned about subtle bugs like https://github.com/cockroachdb/cockroach/pull/114365 and whether our current visibility would allow us to reason them.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver: better comments for on processCallback

This commit updates the comments to better clarify the semantics and the
contract of processCallback.

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152792)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/queue.go` line 800 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3fWO28BKMvBO34BB_:-OZ85e8cAgjJY_E1QzgW:bd99h3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L800)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

I’m calling `onEnqueueResult` with an error even when the returned error is `nil` in certain cases, such as `errReplicaAlreadyInPurgatory` or `errReplicaAlreadyProcessing`. I want to discuss the desired behavior here before making the changes, especially in light of your point below about whether a replica already in the queue should be treated as a success. 

I still want visibility into these cases - at least knowing if a replica is somehow stuck in purgatory or processing. While the replicas are technically in the queue and things might seem fine, I’m concerned about subtle bugs like https://github.com/cockroachdb/cockroach/pull/114365 and whether our current visibility would allow us to reason them.
</blockquote></details>

I see - hadn't noticed the `return nil` with an associated error callback. This makes sense to me. I still think you could take the suggestion, though - in the purgatory and the requeue case, you invoke the callback manually (with a comment). For the other "vanilla" cases that return an error and also want it to be passed to the callback, you use the defer.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver: treat priority update as a success with onEnqueueResult

Previously, bq.addInternal invoked processCallback.onEnqueueResult with a newly
constructed error whenever a replica was already present and re-enqueued with a
higher priority, since the priority had to be passed dynamically. This commit
instead treats the case as a success and passes the updated heap index to
onEnqueueResult, as the caller mainly cares about whether the replica is already
in the queue and its new position.

wenyihu6 committed:
kvserver: rename processCallback processCallback to cb processCallback

Previously, the variable name processCallback shadowed its type name,
which was not ideal. This commit renames the variable to cb.

wenyihu6 committed:
kvserver: call cb.onEnqueueResult in defer on errors

Previously, cb.onEnqueueResult was invoked inline before returning errors, which
was less robust and required explicit calls. This commit refactors the code to
invoke onEnqueueResult in a defer statement when returning a non-nil error. Note
that the function may still call cb.onEnqueueResult with non-nil errors even
when no error is returned, since we want visibility into those cases as well.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152792)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/queue.go` line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3ejgs2Mp45HPjQFGD:-OZBLyPM6v9hGpokoZPi:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L132)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

puragtory
</blockquote></details>

Fixed.
___
*[`pkg/kv/kvserver/queue.go` line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3esba1Q4PzoFwEi_M:-OZBLzCw7mWrlvkhZ4fh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L150)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: for temporal consistency, move `onEnqueue` up above `onProcessResult`.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/queue.go` line 163 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3f1iQ8eAMwiF6Hnfn:-OZBM-Sg5eoqUzC_lkaj:b-llg699) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L163)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Here too could you reflect on whether the callback is necessarily invoked at all? I imagine there are cases where it isn't, and it'd be good to call this out.
</blockquote></details>

Done. It’s a bit more nuanced because `addInternal` and `MaybeAddCallback` behave a bit differently when registering callbacks. Added comments on them to clarify the behavior.
___
*[`pkg/kv/kvserver/queue.go` line 645 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3fGAbC_JlrJ3oghio:-OZBNJ1o9pOLwp8pS2-T:ba6kvax) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L645)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

the new variable name shadows the type, which isn't ideal. Maybe just `processCB processCallback`.
</blockquote></details>

Good call - renamed to `cb processCallback` to match the existing convention used with `(replicaItem) func registerCallback`.
___
*[`pkg/kv/kvserver/queue.go` line 703 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3fMWuCWl63vidzMdk:-OZBNd4a7WEWNoI51TDv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L703)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

ditto (maybe just search for `processCallback processCallback` to get all the offenders.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/queue.go` line 800 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3fWO28BKMvBO34BB_:-OZBNeHcBf3jiKepaWyG:b-3w7iz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L800)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I see - hadn't noticed the `return nil` with an associated error callback. This makes sense to me. I still think you could take the suggestion, though - in the purgatory and the requeue case, you invoke the callback manually (with a comment). For the other "vanilla" cases that return an error and also want it to be passed to the callback, you use the defer.
</blockquote></details>

Makes sense, fixed in a follow up commit. 
___
*[`pkg/kv/kvserver/queue.go` line 874 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3gT5N02Jlv7l1ohKw:-OZBNlvX61N6gp6rdSci:b-f5scs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L874)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Adding an extra anchor comment here as a reminder to do this, since I reviewed the shrinking size code already. Likely best to stack this PR on top.

There is also the case in `SetQueueSize` that would need to call callbacks, so most likely a good idea to factor out a single method that is called from both places to do the callback handling.
</blockquote></details>

I haven’t added the helper function yet, as I’m still considering how to structure it. Invoking the callback inside `bq.removeFromReplicaSetLocked` seems like the natural approach. This would also resolve the semantic issues as well, but it introduces a behavior change by calling the callback while holding a lock. The comment also seems to suggest that there may be things I'm overlooking: 
https://github.com/cockroachdb/cockroach/blob/84697f3097d4a65a0865248f32d63c5eb3ca304d/pkg/kv/kvserver/queue.go#L825-L827
___
*[`pkg/kv/kvserver/replica.go` line 2968 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3gsDpCBTq6t5y5AIG:-OZBP-ml43Ezsu5qnZam:baxfemk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/replica.go#L2968)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Some rate limiting here would be good here:

var logBudgetEnqueueResult atomic.Int
logBudgetEnqueueResult.Store(15)

// in the callback:
if logBudgetEnqueueResult.Add(-1) > 0 {
  // log.Infof
} else {
  // vmodule
}

So basically each decommission nudger run will produce at most 15 messages for enqueue failures (I didn't name the var very well: it was meant to be for the failure case, we should have one atomic per message type).

Alternatively, and simpler, we could gate all logging behind verbosity if the number of replicas we're nudging is (say) >15. But I like getting some information even in the case in which there are more than that number of replicas, which this atomic-based thing would give us.
</blockquote></details>

Good idea. I’m opening a new PR https://github.com/cockroachdb/cockroach/pull/152885 for this, as there are some design decisions I made that are still open for discussion.
___
*[`pkg/kv/kvserver/replica_backpressure.go` line 214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3eIn40VDZVSLZ0oLp:-OZBPJxF7rGFYotEvR7i:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa26dc3b2812e6af8e8c1e418d3a011082278d7c/pkg/kv/kvserver/replica_backpressure.go#L214)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

this TODO seems out of place, and along with the unrelated comment below it tripped me up a little. Consider removing or moving the TODO.
</blockquote></details>

Removed.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
fixup! kvserver: treat priority update as a success with onEnqueueResult

tbg reviewed (approved):
> @tbg reviewed 2 of 2 files at r5, 1 of 1 files at r6, 2 of 2 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152792)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/queue.go` line 874 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152792#-OZ3gT5N02Jlv7l1ohKw:-OZDnaCd4RtEv_rDDC1b:b-vy2w38) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcb8b75c2aefe5c00fd5bf28f9418662e6011/pkg/kv/kvserver/queue.go#L874)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

I haven’t added the helper function yet, as I’m still considering how to structure it. Invoking the callback inside `bq.removeFromReplicaSetLocked` seems like the natural approach. This would also resolve the semantic issues as well, but it introduces a behavior change by calling the callback while holding a lock. The comment also seems to suggest that there may be things I'm overlooking: 
https://github.com/cockroachdb/cockroach/blob/84697f3097d4a65a0865248f32d63c5eb3ca304d/pkg/kv/kvserver/queue.go#L825-L827
</blockquote></details>

Yep, we can revisit separately. I think that comment is mostly FUD, btw. At this point you have an at least equal understanding to Dan (who wrote the comment at the time). You can stack your attempt as a new PR, then we can focus on not messing anything up.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR!

bors r=tbg
